main/core.js

```diff
@@ -39,86 +37,62 @@ async function setup (/** @type {Context} */ ctx) {
     if (filePath) {
-      await fs.writeFile(filePath, await core.logs.get())
+      await fs.writeFile(filePath, logs.get())
     }
   }
   await maybeMigrateFiles()
-  subscribeWithRetry(ctx, core).catch(console.error)
-  await start(core)
-}
-
-async function subscribeWithRetry (
-  /** @type {Context} */ ctx,
-  /** @type {Core} */ core
-) {
-  while (true) {
-    const controller = new AbortController()
-    try {
-      await subscribe(ctx, core, controller.signal)
-    } catch (err) {
-      controller.abort()
-      console.error(err)
-      await timers.setTimeout(1000)
-    }
-  }
-}
-
-async function subscribe (
-  /** @type {Context} */ ctx,
-  /** @type {Core} */ core,
-  /** @type {AbortSignal} */ signal
-) {
-  await Promise.all([
-    (async () => {
-      for await (const metrics of core.metrics.follow({ signal })) {
-        ctx.setTotalJobsCompleted(metrics.totalJobsCompleted)
-      }
-    })(),
-    (async () => {
-      const it = core.activity.follow({ nLines: 0, signal })
-      for await (const activity of it) {
-        ctx.recordActivity(activity)
-        detectChangeInOnlineStatus(activity)
-      }
-    })()
-  ])
+  await start(ctx)
 }
 
 /**
- * @param {Activity} activity
+ * @param {Context} ctx
  */
-function detectChangeInOnlineStatus (activity) {
-  if (
-    activity.type === 'info' &&
-    activity.message.includes('Saturn Node is online')
-  ) {
-    online = true
-  } else if (
-    activity.message === 'Saturn Node started.' ||
-    activity.message.includes('was able to connect') ||
-    activity.message.includes('will try to connect')
-  ) {
-    online = false
-  }
-}
-
-/**
- * @param {Core} core
- */
-async function start (core) {
+async function start (ctx) {
   assert(wallet.getAddress(), 'Core requires FIL address')
   console.log('Starting Core...')
 
-  const childProcess = fork(corePath, {
+  const childProcess = fork(corePath, ['--json'], {
     env: {
       ...process.env,
       FIL_WALLET_ADDRESS: wallet.getAddress(),
       CACHE_ROOT: consts.CACHE_ROOT,
       STATE_ROOT: consts.STATE_ROOT,
       DEPLOYMENT_TYPE: 'station-desktop'
+    },
+    stdio: ['pipe', 'pipe', 'pipe', 'ipc']
+  })
+
+  assert(childProcess.stdout)
+  childProcess.stdout.setEncoding('utf8')
+  childProcess.stdout.on('data', chunk => {
+    logs.push(chunk)
+    for (const line of chunk.split('\n').filter(Boolean)) {
+      const event = JSON.parse(line)
+      switch (event.type) {
+        case 'jobs-completed':
+          ctx.setTotalJobsCompleted(event.total)
+          break
+        case 'activity:info':
+        case 'activity:error': {
+          const activity = {
+            ...event,
+            type: event.type.replace('activity:', ''),
+            timestamp: new Date(),
+            id: randomUUID()
+          }
+          activities.push(ctx, activity)
+          break
+        }
+        default:
+          throw new Error(`Unknown event type: ${event.type}`)
```

Comment:
I think this is going to crash the entire process on unhandled error, is that correct? That seems to be too strict to me. Have you considered treating this edge case as a warning only? I.e. log this error to console.error and send it to Sentry.

Reply:
Good idea! https://github.com/filecoin-station/desktop/pull/730